Rewrite the assembly conflict resolution doc to describe current behavior (#32221)
dotnet/runtime: docs/design/features/assembly-conflict-resolution.md

#### Probe Ordering
The probing locations consists of:

> 1.  Serving Location

Reviewer: ```suggestion
1.  Servicing Location
```

Per #32272

#### Probe Ordering
The probing locations consists of:
1.  Serving Location
1.  App directory

> 1.  Framework directory(s) from higher to lower

Reviewer 2: The doc is no longer self-consistent unforetunately
In the section marked 'Algorithm' the text refers to app being after framework.
In the section marked 'Changes for 2.1+' it is refers back to this list with the expectation that framework is item 3 and app is item 4.

Author: Good catch. I rewrote the doc almost fully to only describe the current behavior - it's not very useful to have a "proposal" doc around since some pieces of it were never implemented and it's hard to read to determine the current behavior.

#### Algorithm
1. Determine the probing paths
1. For each entry in the app's `.deps.json`
1. ->If there's already a resolved entry with the same asset name, then check the assembly and file version of the new entry against the already resolved one. If the new entry is equal or lower, skip it. Otherwise remove the existing one and go to the "else" branch below
1. ->Else (new asset, or replaced with higher version) probe for the actual asset file
1. -->For each probing path except frameworks that are higher-level

> 1. --->If the probing path is a framework, then check its `.deps.json` to see if it contains the exact package (by name and version). If it is, then use that location and end probing for this entry.

Reviewer: ```suggestion
1. --->If the probing path is a framework, then check its `.deps.json` to see if it contains the exact package (by name and version). If it does, then use that location and end probing for this entry.
```

1. --->If the probing path is not a framework, then use that location and end probing for this entry
1. Read the additional deps from `--additional-deps` and repeat steps 3-7
1. Read each framework's deps.json (higher to lower) and repeat steps 3-7
1. Pass the set of assemblies and their paths to the CLR

> Note that for an app both it's `.deps.json` as well as probing path comes before the framework's so mostly it will win. Also because the app will likely reference an OOB package that the framework doesn't (because a framework, at least Microsoft.NETCore.App, has its own metapackage and does not reference OOB packages), the framework probing path never matches up in step 4 for the app's `.deps.json` package\assembly entry, so the "app wins".

Reviewer: ```suggestion
Note that for an app both its `.deps.json` as well as probing path comes before the framework's `.deps.json`, so the app will usually win. Also because the app will likely reference an OOB package that the framework doesn't (because a framework, at least Microsoft.NETCore.App, has its own metapackage and does not reference OOB packages), the framework probing path never matches up in step 4 for the app's `.deps.json` package\assembly entry, so the "app wins".
```

Reviewer: And I think `step 4` -> `step 6`

> The reason for only probing paths from equal or lower level framework in step 3 is that a given framework should never have a dependency on a higher-level framework, and is expected to find deps assets in its layer or lower. This is also required so that the given framework can find its asset, and replace the higher-level asset (see next paragraph).

Reviewer: Is `step 3` -> `step 5`?
